AutoGraphTests/WebSocketClientTests.swift

```diff
@@ -0,0 +1,349 @@
+        self.subject = try! MockWebSocketClient(url: url, webSocket: self.webSocket)
+        self.webSocketDelegate = MockWebSocketClientDelegate()
+        self.subject.delegate = self.webSocketDelegate
+    }
+    
+    override func tearDown() {
+        self.subject = nil
+        self.webSocket = nil
+        self.webSocketDelegate = nil
+        super.tearDown()
+    }
+    
+    func testSendSubscriptionResponseHandlerIsCalledOnSuccess() {
+        var called = false
+        let request = try! SubscriptionRequest(request: FilmSubscriptionRequest(), operationName: "film")
+        _ = self.subject.subscribe(request: request, responseHandler: SubscriptionResponseHandler(completion: { result in
+            
+            guard case .success(_) = result else {
+                XCTFail()
+                return
+            }
+            
+            called = true
+        }))
+        
+        waitFor(delay: kDelay)
+        XCTAssertTrue(called)
+    }
+    
+    
+    func testSendSubscriptionCorrectlyDecodesResponse() {
+        var film: Film?
+        let request = try! SubscriptionRequest(request: FilmSubscriptionRequest(), operationName: "film")
+        _ = self.subject.subscribe(request: request, responseHandler: SubscriptionResponseHandler(completion: { result in
+            
+            guard case let .success(data) = result else {
+                XCTFail()
+                return
+            }
+            
+            film = try! JSONDecoder().decode(Film.self, from: data)
+        }))
+        
+        XCTAssertNotNil(film)
+        XCTAssertEqual(film?.remoteId, "ZmlsbXM6MQ==")
+    }
+    
+    func testUnsubscribeRemovesSubscriptions() {
+        let request = try! SubscriptionRequest(request: FilmSubscriptionRequest(), operationName: "film")
+        let subscriber = self.subject.subscribe(request: request, responseHandler: SubscriptionResponseHandler(completion: { _ in }))
+        
+        XCTAssertTrue(self.subject.subscriptions["film}"]!.contains(where: {$0.key == subscriber }))
+        
+        try! self.subject.unsubscribe(subscriber: subscriber)
+        
+        XCTAssertEqual(self.subject.subscriptions.count, 0)
+    }
+    
+    
+    func testUnsubscribeAllRemovesSubscriptions() {
+        let request = try! SubscriptionRequest(request: FilmSubscriptionRequest(), operationName: "film")
+        _ = self.subject.subscribe(request: request, responseHandler: SubscriptionResponseHandler(completion: { _ in }))
+
+        let request2 = try! SubscriptionRequest(request: FilmSubscriptionRequest(), operationName: "film1")
+        _ = self.subject.subscribe(request: request2, responseHandler: SubscriptionResponseHandler(completion: { _ in }))
+
+        try! self.subject.unsubscribeAll(request: request)
+
+        XCTAssertEqual(self.subject.subscriptions.count, 1)
+        
+        try! self.subject.unsubscribeAll(request: request2)
+        
+        XCTAssertEqual(self.subject.subscriptions.count, 0)
```

Comment:
we should have more than 1 of these subscriptions for the same set on one of these so we see it go from 2 to 0